### PR TITLE
[ENH] Add support for endpoint that will return `nb:Diagnosis` options from SNOMED

### LINKED ISSUE #211: Update vocabulary path to parse and return `nb:Diagnosis` options
Add workflow for the path `/attributes/nb:Diagnosis/vocab` that:
- [x] Reads in a file of SNOMED diagnosis term-label mappings (downloaded from Athena) on startup + creates a lookup file
  - a copy should probably be stored in the repo for now, under https://github.com/neurobagel/api/tree/main/vocab/backup_external
  - The diagnosis term-label mappings can be found here https://github.com/neurobagel/annotation_tool/blob/main/static/diagnosisTerms.json
- [x] Processes the mappings into a `VocabLabelsResponse` object
- [x] append the missing terms (see below) to the SNOMED-CT .json dump
  - already done in .json used by annotation tool
- [x] update the code snippet in https://github.com/neurobagel/planning/issues/52 to also include the appended terms that are missing, so it is re-executable
- ~Allows only password-authenticated users to access that endpoint~

Other TODO:
- [x] wrap requests to external vocab APIs in a `try-except` for httpx network errors (currently, if the network connection cannot be established if, e.g. the internet is down, an exception is raised and the app cannot start)
  - [x] if catch is hit, raise informative warning and use local copy of vocab file

### PR DESCRIPTION
**Pull request by alyssadai**
- Closes #211 

Changes proposed in this pull request:

- Added JSON dump of SNOMED CT disorder terms
- On startup, app will now read in the file of SNOMED diagnosis terms and store a lookup file where entries are in the format `<id>:<name>`
- Implemented logic for `/attributes/nb:Diagnosis/vocab` which loads + processes the lookup file contents into a `VocabLabelsResponse` object
- Added graceful exception handling for network errors encountered when requesting external vocabs on startup (cogatlas)

## Checklist

- [x] PR has an interpretable title with a prefix (`[ENH]`, `[FIX]`, `[REF]`, `[TST]`, `[CI]`, `[MNT]`, `[INF]`, `[MODEL]`, `[DOC]`) _(see https://neurobagel.org/contributing/pull_requests for more info)_
- [x] PR links to GitHub issue with mention `Closes #XXXX`
- [x] Tests pass
- [x] Checks pass

For new features:
- [x] Tests have been added

For bug fixes:
- [ ] There is at least one test that would fail under the original bug conditions.

### CODE REVIEW

**Review by rmanaem — APPROVED** [review]
Thanks for the PR @alyssadai! Everything looks good. I just left a suggestion and a question, please take a look and then :cook:

File: app/api/crud.py [hunk 1]
```diff
@@ -275,11 +275,11 @@ async def get_controlled_term_attributes() -> list:
     return results_list
 
 
-async def get_term_labels_for_cogatlas(
-    term_labels_path: Path,
+async def get_term_labels_for_vocab(
+    term_labels_path: Path, vocabulary_name: str, namespace_prefix: str
```
Comment (surchs) [inline app/api/crud.py:279]: :tada:

File: app/api/routers/attributes.py [hunk 2]
```diff
@@ -12,10 +12,19 @@ async def get_term_labels_for_vocab(
...
-    if data_element_URI is DataElementURI.assessment:
-        return await crud.get_term_labels_for_cogatlas(
-            term_labels_path=request.app.state.cogatlas_term_lookup_path
+    if data_element_URI == DataElementURI.assessment:
+        return await crud.get_term_labels_for_vocab(
+            term_labels_path=request.app.state.cogatlas_term_lookup_path,
+            vocabulary_name="Cognitive Atlas Tasks",
+            namespace_prefix="cogatlas",
         )
+    else:
```
Comment (surchs) [inline app/api/routers/attributes.py:21]: I think this `else` is not needed here.

It's a (relatively) new language feature, but you could also take a look at https://peps.python.org/pep-0634/#the-match-statement

File: app/api/routers/attributes.py [hunk 3]
```diff
@@ -30,6 +39,9 @@ async def get_terms(
 
     if data_element_URI == DataElementURI.assessment:
         term_labels_path = request.app.state.cogatlas_term_lookup_path
+    else:
```
Comment (surchs) [inline app/api/routers/attributes.py:42]: is there are reason you add the `else` here?

**Review by surchs — COMMENTED** [review]
I somehow missed the previous review. Just adding my comments here in case it helps

File: app/api/routers/attributes.py (hunk 3 above)
Comment (alyssadai) [reply app/api/routers/attributes.py:42]: Hm, I think you're right that it's not actually needed!

File: app/api/utility.py [hunk 4]
```diff
@@ -369,26 +369,38 @@ def fetch_and_save_cogatlas(output_path: Path):
...
+                """
+            )
+            # Use backup copy of the raw vocabulary JSON
+            vocab = load_json(BACKUP_VOCAB_DIR / "cogatlas_task.json")
+    except httpx.NetworkError as exc:
         warnings.warn(
-            f"""
-            The API was unable to fetch the Cognitive Atlas task vocabulary (https://www.cognitiveatlas.org/tasks/a/) from the source and will default to using a local backup copy of the vocabulary instead.
+            f""""
+            Fetching of the Cognitive Atlas task vocabulary (https://www.cognitiveatlas.org/tasks/a/) from the source failed due to a network error.
```
[the remark this answers is not in the document]
Comment (alyssadai) [reply app/api/utility.py:400]: I don't think so! `NetworkError` falls under the `RequestError` umbrella, which is separate from `HTTPStatusError` (which is what's raised by the `raise_for_status` call suggested by your other comment; see https://www.python-httpx.org/quickstart/#exceptions / https://www.python-httpx.org/exceptions/#exception-classes). `NetworkError` specifically includes problems establishing a network connection, so this `except` allows the API to still start up even if the user isn't connected to the internet. `raise_for_status`/`HTTPStatusError` covers cases where the request has gone through, whereas to my understanding, when `NetworkError`s arise, there's no response attribute (and so no status code).

That said, I can try to rewrite this bit to except _both_ `HTTPStatusError` (using `raise_for_status()`) and `NetworkError`. I think it should give me the same behaviour as the code currently.

[the remark this answers is not in the document]
Comment (alyssadai) [reply app/api/utility.py:400]: So after some testing, I think I will keep the current syntax for now as switching to using `raise_for_status()` also requires more complicated mocking of the `httpx` functions in tests, which I think is not necessarily worthwhile for now as checking the status code directly serves the same purpose.

Happy to revisit this in the future if we need more complex exception handling/flow control!

**Pull request merged**